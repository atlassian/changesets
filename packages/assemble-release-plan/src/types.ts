import {
  VersionType,
  PreState,
  NewChangeset,
  Config,
  ReleasePlan
} from "@changesets/types";
import { Packages } from "@manypkg/get-packages";

export type InternalRelease = {
  name: string;
  type: VersionType;
  oldVersion: string;
  changesets: string[];
};

export type PreInfo = {
  state: PreState;
  preVersions: Map<string, number>;
};

export type AssembleReleasePlan = (
  changesets: NewChangeset[],
  packages: Packages,
  config: Config,
  preState: PreState | undefined,
  snapshot?: string | boolean | undefined
) => ReleasePlan;
