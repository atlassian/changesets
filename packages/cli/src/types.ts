export type CliOptions = {
  sinceMaster?: boolean;
  verbose?: boolean;
  output?: string;
  otp?: string;
  empty?: boolean;
  since?: string;
  ignore?: string | string[];
  snapshot?: string | boolean;
  snapshotPrereleaseTemplate?: string;
  tag?: string;
  gitTag?: boolean;
  open?: boolean;
  major?: string | string[];
  minor?: string | string[];
  patch?: string | string[];
  message?: string;
};

export type CommandOptions = CliOptions & {
  cwd: string;
};
