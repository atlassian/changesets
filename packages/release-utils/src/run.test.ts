import fixturez from "fixturez";
import fs from "fs-extra";
import path from "path";
import writeChangeset from "@changesets/write";
import { Changeset } from "@changesets/types";
import { runVersion } from "./run";
import { add, commit } from "@changesets/git";
import spawn from "spawndamnit";
import fileUrl from "file-url";

let f = fixturez(__dirname);

const linkNodeModules = async (cwd: string) => {
  await fs.symlink(
    path.join(__dirname, "..", "..", "..", "node_modules"),
    path.join(cwd, "node_modules")
  );
};
const writeChangesets = (changesets: Changeset[], cwd: string) => {
  return Promise.all(changesets.map(commit => writeChangeset(commit, cwd)));
};

beforeEach(() => {
  jest.clearAllMocks();
});

async function setupRepoAndClone(cwd: string) {
  await spawn("git", ["init"], { cwd });
  await add(".", cwd);
  await commit("commit1", cwd);

  // Make a 1-commit-deep shallow clone of this repo
  let clone = f.temp();
  await spawn(
    "git",
    // Note: a file:// URL is needed in order to make a shallow clone of
    // a local repo
    ["clone", "--depth", "1", fileUrl(cwd), "."],
    {
      cwd: clone
    }
  );
  await spawn("git", ["checkout", "-b", "some-other-branch"], { cwd });
  return clone;
}

describe("version", () => {
  it.only("creates simple PR", async () => {
    let cwd = f.copy("simple-project");

    await writeChangesets(
      [
        {
          releases: [
            {
              name: "pkg-a",
              type: "minor"
            },
            {
              name: "pkg-b",
              type: "minor"
            }
          ],
          summary: "Awesome feature"
        }
      ],
      cwd
    );

    const clone = await setupRepoAndClone(cwd);

    await linkNodeModules(clone);

    await runVersion({
      cwd: clone
    });

    await spawn("git", ["checkout", "changeset-release/master"], { cwd });

    expect(
      await fs.readFile(
        path.join(cwd, "packages", "pkg-a", "package.json"),
        "utf-8"
      )
    ).toMatchInlineSnapshot(`
      "{
        \\"name\\": \\"pkg-a\\",
        \\"version\\": \\"1.1.0\\",
        \\"dependencies\\": {
          \\"pkg-b\\": \\"1.1.0\\"
        }
      }
      "
    `);

    expect(
      await fs.readFile(
        path.join(cwd, "packages", "pkg-b", "package.json"),
        "utf-8"
      )
    ).toMatchInlineSnapshot(`
      "{
        \\"name\\": \\"pkg-b\\",
        \\"version\\": \\"1.1.0\\"
      }
      "
    `);

    expect(
      await fs.readFile(
        path.join(cwd, "packages", "pkg-a", "CHANGELOG.md"),
        "utf-8"
      )
    ).toMatchInlineSnapshot(`
      "# pkg-a

      ## 1.1.0
      ### Minor Changes

      - fdf577e: Awesome feature

      ### Patch Changes

      - Updated dependencies [fdf577e]
        - pkg-b@1.1.0
      "
    `);
    expect(
      await fs.readFile(
        path.join(cwd, "packages", "pkg-b", "CHANGELOG.md"),
        "utf-8"
      )
    ).toMatchInlineSnapshot(`
      "# pkg-b

      ## 1.1.0
      ### Minor Changes

      - fdf577e: Awesome feature
      "
    `);
  });

  it.only("only includes bumped packages in the PR body", async () => {
    let cwd = f.copy("simple-project");

    await writeChangesets(
      [
        {
          releases: [
            {
              name: "pkg-a",
              type: "minor"
            }
          ],
          summary: "Awesome feature"
        }
      ],
      cwd
    );

    const clone = await setupRepoAndClone(cwd);

    await linkNodeModules(clone);

    await runVersion({
      cwd: clone
    });

    await spawn("git", ["checkout", "changeset-release/master"], { cwd });

    expect(
      await fs.readFile(
        path.join(cwd, "packages", "pkg-a", "package.json"),
        "utf-8"
      )
    ).toMatchInlineSnapshot(`
      "{
        \\"name\\": \\"pkg-a\\",
        \\"version\\": \\"1.1.0\\",
        \\"dependencies\\": {
          \\"pkg-b\\": \\"1.0.0\\"
        }
      }
      "
    `);

    expect(
      await fs.readFile(
        path.join(cwd, "packages", "pkg-b", "package.json"),
        "utf-8"
      )
    ).toMatchInlineSnapshot(`
      "{
        \\"name\\": \\"pkg-b\\",
        \\"version\\": \\"1.0.0\\"
      }
      "
    `);

    expect(
      await fs.readFile(
        path.join(cwd, "packages", "pkg-a", "CHANGELOG.md"),
        "utf-8"
      )
    ).toMatchInlineSnapshot(`
      "# pkg-a

      ## 1.1.0
      ### Minor Changes

      - d8d2f4f: Awesome feature
      "
    `);
    await expect(
      fs.readFile(path.join(cwd, "packages", "pkg-b", "CHANGELOG.md"), "utf-8")
    ).rejects.toMatchObject({ code: "ENOENT" });
  });

  it("doesn't include ignored package that got a dependency update in the PR body", async () => {
    let cwd = f.copy("ignored-package");

    await writeChangesets(
      [
        {
          releases: [
            {
              name: "ignored-package-pkg-b",
              type: "minor"
            }
          ],
          summary: "Awesome feature"
        }
      ],
      cwd
    );

    const clone = await setupRepoAndClone(cwd);

    await linkNodeModules(clone);

    await runVersion({
      cwd
    });
  });
});
