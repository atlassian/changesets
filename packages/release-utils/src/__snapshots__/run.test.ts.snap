// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`version doesn't include ignored package that got a dependency update in the PR body 1`] = `
Array [
  Object {
    "base": "some-branch",
    "body": "This PR was opened by the [Changesets release](https://github.com/changesets/action) GitHub action. When you're ready to do a release, you can merge this and publish to npm yourself or [setup this action to publish automatically](https://github.com/changesets/action#with-publishing). If you're not ready to do a release yet, that's fine, whenever you add more changesets to some-branch, this PR will be updated.

# Releases
## ignored-package-pkg-b@1.1.0

### Minor Changes

-   Awesome feature
",
    "head": "changeset-release/some-branch",
    "owner": "changesets",
    "repo": "action",
    "title": "Version Packages",
  },
]
`;
